File: sdk/clojure/test/bsdkrun/types_test.clj
(ns bsdkrun.types-test
  (:require [clojure.test :refer [deftest is]]
            [bsdkrun.types :as types]))

(deftest sandbox-info-from-row-full
  (let [info (types/sandbox-info-from-row
              {"id" "abc123" "name" "api" "image" "alpine" "kind" "linux"
               "command" "sleep 300" "running" true "exit_code" nil "pid" 42
               "detached" true "cpus" 2 "mem" 512 "volume" nil
               "state_dir" "/s" "network" "devnet" "net_ip" "10.0.0.2"
               "created_at" 1700000000 "finished_at" nil
               "ports" [{"bind" "127.0.0.1" "host" 8080 "guest" 80}]})]
    (is (= "abc123" (:id info)))
    (is (= "api" (:name info)))
    (is (= "running" (:status info)))
    (is (true? (:running info)))
    (is (nil? (:exit-code info)))
    (is (= 42 (:pid info)))
    (is (true? (:detached info)))
    (is (= 2 (:cpus info)))
    (is (= 512 (:mem info)))
    (is (= "/s" (:state-dir info)))
    (is (= "devnet" (:network info)))
    (is (= "10.0.0.2" (:net-ip info)))
    (is (= 1700000000 (:created-at info)))
    (is (nil? (:finished-at info)))
    (is (= 1 (count (:ports info))))
    (is (= {:bind "127.0.0.1" :host 8080 :guest 80} (first (:ports info))))))

(deftest sandbox-info-from-row-sparse
  (let [info (types/sandbox-info-from-row {"id" "abc123"})]
    (is (= "abc123" (:id info)))
    (is (= "" (:image info)))
    (is (= "exited" (:status info)))
    (is (false? (:running info)))
    (is (nil? (:exit-code info)))
    (is (nil? (:pid info)))
    (is (= 0 (:cpus info)))
    (is (= 0 (:mem info)))
    (is (= [] (:ports info)))))

(deftest port-forward-from-row
  (is (= {:bind "127.0.0.1" :host 8080 :guest 80}
         (types/port-forward-from-row {"bind" "127.0.0.1" "host" 8080 "guest" 80}))))

(deftest image-info-from-row
  (is (= {:id "i1" :reference "alpine:3.19" :digest "sha256:x" :size 1024
          :rootfs "/r" :created-at 111}
         (types/image-info-from-row
          {"id" "i1" "reference" "alpine:3.19" "digest" "sha256:x"
           "size" 1024 "rootfs" "/r" "created_at" 111}))))

(deftest volume-info-from-row
  (is (= {:name "web" :guest nil :base nil :path "/v/web" :size "10G"
          :created-at 222 :tracked true}
         (types/volume-info-from-row
          {"name" "web" "path" "/v/web" "size" "10G" "created_at" 222
           "tracked" true}))))

(deftest volume-info-from-row-nil-created-at
  (is (nil? (:created-at (types/volume-info-from-row {"name" "web" "path" "/v/web"})))))

(deftest network-info-from-row
  (is (= {:name "devnet" :subnet "10.0.0.0/24" :gateway "10.0.0.1"
          :members 2 :running 1 :up true :created-at 333}
         (types/network-info-from-row
          {"name" "devnet" "subnet" "10.0.0.0/24" "gateway" "10.0.0.1"
           "members" 2 "running" 1 "up" true "created_at" 333}))))

(deftest network-info-from-row-defaults
  (let [info (types/network-info-from-row {"name" "devnet"})]
    (is (= 0 (:members info)))
    (is (= 0 (:running info)))
    (is (false? (:up info)))
    (is (nil? (:created-at info)))))

(deftest result-helpers
  (let [ok {:stdout "hello\n\n" :stderr "" :exit-code 0 :command "x"}
        bad {:stdout "" :stderr "boom" :exit-code 1 :command "x"}]
    (is (true? (types/ok? ok)))
    (is (= "hello" (types/text ok)))
    (is (false? (types/ok? bad)))
    (is (thrown? clojure.lang.ExceptionInfo (types/throw-if-failed! bad)))
    (is (= ok (types/throw-if-failed! ok)))))

(deftest lines-helper
  (is (= ["a" "b"] (types/lines {:stdout "a\n\nb\n"}))))

(deftest json-helper
  (is (= {"x" 1} (types/json {:stdout "{\"x\":1}"}))))

(deftest read-json-rows
  (is (= [] (types/read-json-rows "")))
  (is (= [] (types/read-json-rows nil)))
  (is (= [{"a" 1}] (types/read-json-rows "[{\"a\":1}]"))))

;; --- GraphQL decoders (bsdkrun.client) ----------------------------------

(deftest sandbox-info-from-graphql-full
  (let [info (types/sandbox-info-from-graphql
              {"id" "abc123" "name" "api" "image" "alpine" "kind" "linux"
               "command" "sleep 300" "status" "running" "running" true
               "exitCode" nil "pid" 42 "detached" true "cpus" 2 "mem" 512
               "volume" nil "stateDir" "/s" "network" "devnet" "netIp" "10.0.0.2"
               "createdAt" 1700000000 "finishedAt" nil
               "ports" [{"bind" "127.0.0.1" "host" 8080 "guest" 80}]})]
    (is (= "abc123" (:id info)))
    (is (= "running" (:status info)))
    (is (true? (:running info)))
    (is (nil? (:exit-code info)))
    (is (= 42 (:pid info)))
    (is (= "/s" (:state-dir info)))
    (is (= "10.0.0.2" (:net-ip info)))
    (is (= 1700000000 (:created-at info)))
    (is (nil? (:finished-at info)))
    (is (= {:bind "127.0.0.1" :host 8080 :guest 80} (first (:ports info))))))

(deftest sandbox-info-from-graphql-sparse
  (let [info (types/sandbox-info-from-graphql {"id" "abc123"})]
    (is (= "abc123" (:id info)))
    (is (= "" (:image info)))
    (is (false? (:running info)))
    (is (nil? (:exit-code info)))
    (is (= 0 (:cpus info)))
    (is (= [] (:ports info)))))

(deftest command-result-from-graphql-defaults-blank-strings
  (is (= {:exit-code 0 :stdout "ok" :stderr ""}
         (types/command-result-from-graphql {"exitCode" 0 "stdout" "ok" "stderr" nil}))))

(deftest shell-session-info-from-graphql
  (is (= {:id "s1" :machine-id "m1" :finished false :truncated true}
         (types/shell-session-info-from-graphql
          {"id" "s1" "machineId" "m1" "finished" false "truncated" true}))))

File: sdk/go/types_test.go
package bsdkrun

import (
	"errors"
	"reflect"
	"testing"
)

const psRow = `[
  {
    "id": "abc123def456",
    "name": "web",
    "image": "alpine",
    "kind": "linux",
    "command": "sleep 300",
    "running": true,
    "exit_code": null,
    "pid": 42,
    "detached": true,
    "cpus": 2,
    "mem": 512,
    "volume": null,
    "state_dir": "/state",
    "network": "devnet",
    "net_ip": "10.88.0.2",
    "ports": [{"bind": "127.0.0.1", "host": 2222, "guest": 22}],
    "created_at": 1700000000,
    "finished_at": null
  }
]`

func TestParseSandboxRows(t *testing.T) {
	rows, err := parseSandboxRows(psRow)
	if err != nil {
		t.Fatal(err)
	}
	if len(rows) != 1 {
		t.Fatalf("got %d rows", len(rows))
	}
	info := rows[0]
	if info.ID != "abc123def456" || info.Name != "web" || info.Kind != "linux" {
		t.Fatalf("bad identity fields: %+v", info)
	}
	if info.Status != "running" || !info.Running {
		t.Fatalf("status not derived: %+v", info)
	}
	if info.PID == nil || *info.PID != 42 {
		t.Fatalf("pid: %+v", info.PID)
	}
	if info.ExitCode != nil || info.FinishedAt != nil {
		t.Fatalf("nullables should be nil: %+v", info)
	}
	if info.CreatedAt != 1700000000 {
		t.Fatalf("created_at: %d", info.CreatedAt)
	}
	want := []PortForward{{Host: 2222, Guest: 22, Bind: "127.0.0.1"}}
	if !reflect.DeepEqual(info.Ports, want) {
		t.Fatalf("ports: %+v", info.Ports)
	}
}

func TestParseSandboxRowsEmpty(t *testing.T) {
	rows, err := parseSandboxRows("")
	if err != nil || len(rows) != 0 {
		t.Fatalf("rows=%v err=%v", rows, err)
	}
}

func TestSandboxInfoFromGraphQL(t *testing.T) {
	// The schema is camelCase; createdAt/finishedAt arrive as
	// decimal-string unix timestamps.
	info := sandboxInfoFromGraphQL(map[string]any{
		"id":         "abc123",
		"name":       nil,
		"image":      "alpine",
		"kind":       "linux",
		"command":    "sleep 1",
		"status":     "running",
		"running":    true,
		"exitCode":   nil,
		"pid":        float64(42),
		"detached":   true,
		"cpus":       float64(2),
		"mem":        float64(512),
		"volume":     nil,
		"stateDir":   "/s",
		"createdAt":  "1700000000",
		"finishedAt": nil,
		"network":    nil,
		"netIp":      nil,
		"ports":      []any{map[string]any{"bind": "127.0.0.1", "host": float64(2222), "guest": float64(22)}},
	})
	if info.ID != "abc123" || info.CreatedAt != 1700000000 {
		t.Fatalf("%+v", info)
	}
	if info.PID == nil || *info.PID != 42 {
		t.Fatalf("pid: %+v", info.PID)
	}
	if info.Status != "running" {
		t.Fatalf("status: %q", info.Status)
	}
	if len(info.Ports) != 1 || info.Ports[0].Host != 2222 || info.Ports[0].Guest != 22 {
		t.Fatalf("ports: %+v", info.Ports)
	}
}

func TestSandboxInfoFromGraphQLDerivesStatus(t *testing.T) {
	info := sandboxInfoFromGraphQL(map[string]any{"id": "x", "running": false})
	if info.Status != "exited" {
		t.Fatalf("status: %q", info.Status)
	}
}

func TestCommandResultFromGraphQL(t *testing.T) {
	result := commandResultFromGraphQL(map[string]any{
		"exitCode": float64(3),
		"stdout":   "out",
		"stderr":   "err",
	})
	if result.ExitCode != 3 || result.Stdout != "out" || result.Stderr != "err" {
		t.Fatalf("%+v", result)
	}
}

func TestShellSessionInfoFromGraphQL(t *testing.T) {
	info := shellSessionInfoFromGraphQL(map[string]any{
		"id":        "sess-1",
		"machineId": "abc",
		"finished":  false,
		"truncated": true,
	})
	if info.ID != "sess-1" || info.MachineID != "abc" || info.Finished || !info.Truncated {
		t.Fatalf("%+v", info)
	}
}

func TestResultHelpers(t *testing.T) {
	res := &Result{Stdout: "a\nb\n\n", Stderr: "", ExitCode: 0, Command: "exec x"}
	if !res.Ok() {
		t.Fatal("Ok")
	}
	if res.Text() != "a\nb" {
		t.Fatalf("Text: %q", res.Text())
	}
	if !reflect.DeepEqual(res.Lines(), []string{"a", "b"}) {
		t.Fatalf("Lines: %v", res.Lines())
	}
	if err := res.Err(); err != nil {
		t.Fatalf("Err: %v", err)
	}

	jsonRes := &Result{Stdout: `{"x": 1}`}
	var parsed map[string]int
	if err := jsonRes.JSON(&parsed); err != nil || parsed["x"] != 1 {
		t.Fatalf("JSON: %v %v", parsed, err)
	}

	failed := &Result{Stdout: "o", Stderr: "e", ExitCode: 3, Command: "exec y"}
	if failed.Ok() {
		t.Fatal("Ok on failure")
	}
	var cmdErr *CommandFailedError
	if err := failed.Err(); !errors.As(err, &cmdErr) {
		t.Fatalf("Err type: %v", err)
	} else if cmdErr.ExitCode != 3 || cmdErr.Stdout != "o" || cmdErr.Stderr != "e" || cmdErr.Command != "exec y" {
		t.Fatalf("%+v", cmdErr)
	}
}

func TestPortForwardString(t *testing.T) {
	if got := (PortForward{Host: 2222, Guest: 22}).String(); got != "2222:22" {
		t.Fatalf("String: %q", got)
	}
}

func TestAsInt(t *testing.T) {
	cases := []struct {
		in   any
		want int64
		ok   bool
	}{
		{float64(7), 7, true},
		{"1700000000", 1700000000, true},
		{" 42 ", 42, true},
		{nil, 0, false},
		{"x", 0, false},
		{true, 0, false},
	}
	for _, c := range cases {
		got, ok := asInt(c.in)
		if got != c.want || ok != c.ok {
			t.Errorf("asInt(%v) = %d,%v; want %d,%v", c.in, got, ok, c.want, c.ok)
		}
	}
}
